import socket
import argparse
import sys
import getpass


def login(sock):
    print(client_listen(sock), end="")
    msg = "USER {}".format(input("USER : "))
    client_sendcmd(sock, msg)
    print(client_listen(sock), end="")
    msg = "PASS {}".format(getpass.getpass("PASS : "))
    client_sendcmd(sock, msg)
    print(client_listen(sock), end="")

def client_sendcmd(sock, msg):
    msg = "{}\n".format(msg)
    sock.send(msg.encode())

def client_listen(sock, to=0.5):
    response = ""
    while True:
        try:
            sock.settimeout(to)
            data = sock.recv(1024)
        except Exception as e:
            data = ''
        finally:
            if not data:
                break
            else:
                response += data.decode("utf-8")

    return response

def init_data_port(sock, sock_data=None):
    local_infos = sock.getsockname()
    if not sock_data:
        sock_data = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock_data.bind((local_infos[0], 1025))
        sock_data.listen(1)
    client_sendcmd(sock, "PORT 172,16,58,1,4,1")
    return sock_data

def ftp_connect(host):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect(host)
   
    login(sock)
    sock_data = init_data_port(sock)
    while True:
        msg = input()
        client_sendcmd(sock, msg)
        rep = client_listen(sock)
        print(rep, end="")

        if '150' in rep:
            sock_f, addr = sock_data.accept()
            print(client_listen(sock_f, None))
            sock_data = init_data_port(sock, sock_data)

        if '221' in rep:
            break

    sock.close()


def get_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--ip", help="Ip of the ftp server", type=str, default="127.0.0.1")
    parser.add_argument("-p", "--port", help="Port of the ftp server", type=int, default=21)
    return parser.parse_args()


if __name__ == "__main__":
    args = get_arguments() 

    #try:
    ftp_connect((args.ip, args.port))
    #except Exception as e:
    #    print("Error : {}".format(e))
